module/VuFind/src/VuFind/Record/FallbackLoader/Summon.php

```diff
@@ -105,7 +105,7 @@ protected function fetchSingleRecord($id)
                 $params = new ParamBag(
                     ['summonIdType' => Connector::IDENTIFIER_BOOKMARK]
                 );
-                return $this->backend->retrieve($bookmark, $params);
+                return $this->searchService->retrieve('Summon', $bookmark, $params);
```

Comment:
I've just realized we should only stop using the deprecated service method calls in 9.0, because there might be code relying on the old style events. So even new code should use them until 9.0 instead of commands. Sorry about causing confusion here.

Reply:
And if we wanted to be truly fully compatible even with new commands (not replacing deprecated `Service` methods) I guess they too should trigger legacy style events until 9.0...?

Reply:
Ahh, thanks for remembering. That also applies to my work in #2037 -- I've just committed a small fix (83216014509000dca39ccd27acd883aeab8ed9ac) to make sure the sitemap generator still uses the deprecated method (for now).

Reply:
I don't think it's necessary to trigger legacy-style events for new commands; in 8.0 we are announcing that people should be updating code to use the new style of events in preparation for 9.0. If they are writing new code for new events, they should use the new style, not the deprecated style. Or am I missing a larger implication?

Reply:
The event target changes, which unfortunately cannot be avoided if we want to remove the possibility of direct backend access.

All `VuFind\Search` event listeners receive all events, and if they attempt to call a `BackendInterface` method on the target, that would cause an exception with the new events, I believe. Not sure how likely is is to have this kind of code out there. At least most of our code follows the pattern

```php
$backend = $event->getTarget();
if ($backend === $this->backend) {
```

which would not be a problem even if the target is a command object.

Reply:
Yes, that makes sense. My gut feeling is that the risk here is relatively low, and it's probably better to deal with problems as they arise rather than do a lot of work that we'll almost immediately have to undo anyway... but if you feel otherwise (or if others do), please let me know.

Reply:
I agree, and in any case 8.0 is a major version change so breakage is allowed, even if we are trying our best to avoid it until 9.0. Maybe a short note regarding possible code that handles the target as a `BackendInterface` instance could be added to the release notes?

Reply:
And I guess there might also be breakage if the code expects that some parameters exist, so maybe the note should be something along the lines that "much effort has been made to ensure that in most cases there is no breakage, but best to check your listener code just to be sure".

Reply:
Here is the current changelog note:

The VuFindSearch\Service has been refactored to use the Command Pattern; you can now send command objects to the service's invoke method instead of calling more specific search/retrieve/etc. methods. These command objects are also shared with search listeners. The command system makes it easier to customize search behavior without having to rewrite the search service itself. This has been implemented in a mostly-backward-compatible way, but the old way of doing things is deprecated in this release and will be removed in VuFind 9.0. If you make calls to the search service in your code, and/or if you have written any custom search listeners, you should update your code to use command objects instead of the old techniques. When using command objects, some event details will be different (some arguments may not be present, and the event target will also be different). See pull request #1967 for details and examples.

Feel free to edit this directly if you want to improve it.

I also think it may be a good idea to include some content about the command pattern refactoring in this year's VuFind Summit, to make sure the community is aware of this and so that we have a recording we can share in the future. I'd be happy to collaborate with you on such a thing if you might be interested.

Reply:
The note looks perfect to me. Will get back to you on Slack re VuFind Summit.